### PR TITLE
Refactor to save / print shmem usage stats using common struct.

### PR DESCRIPTION
**Pull request by gapisback**
This commit refactors shared memory usage stats fields to drive-off shminfo_usage_stats{} struct entirely. Add platform_save_usage_stats(), used by platform_shm_print_usage().

This refactoring paves the way for upcoming PR #569 which is adding more memory-usage stats fields.

-----
### NOTE To the reviewer:

There is zero logic change here. It's mainly a data structural change, rearranging where the memory-usage stats fields are tracked. They now belong to a common struct, so most of the diffs are to access the new `usage.<field>` reference.

Minor enhancements in the memory usage stats tracking / printing code is done, for improved understanding of memory usage metrics.

### CODE REVIEW

File: src/platform_linux/shmem.c [hunk 1]
```diff
@@ -104,20 +128,10 @@ typedef struct shmem_heap {
...
-   uint32 shm_nlarge_frags_inuse;
-   uint32 shm_nlarge_frags_inuse_HWM;
-   size_t shm_used_by_large_frags; // Actually reserved
-   size_t shm_nfrees;              // # of calls to free memory
-   size_t shm_nfrees_last_frag;    // Freed last small-fragment
 
-   size_t shm_total_bytes; // Total size of shared segment allocated initially.
-   size_t shm_free_bytes;  // Free bytes of memory left (that can be allocated)
-   size_t shm_used_bytes;  // Used bytes of memory left (that were allocated)
-   size_t shm_used_bytes_HWM; // High-water mark of memory used bytes
```
Comment (gapisback) [inline src/platform_linux/shmem.c:118]: This set of deleted stats-fields are now dropped into common `shminfo_usage_stats{}` struct; listed below.

File: src/platform_linux/shmem.c [hunk 2]
```diff
@@ -361,19 +393,17 @@ platform_shmcreate(size_t            size,
...
    shmem_heap *shm = (shmem_heap *)shmaddr;
 
-   shm->shm_start          = shmaddr;
-   shm->shm_end            = (shmaddr + size);
-   shm->shm_next           = (shmaddr + sizeof(shmem_heap));
-   shm->shm_total_bytes    = size;
-   free_bytes              = (size - sizeof(shmem_heap));
-   shm->shm_free_bytes     = free_bytes;
-   shm->shm_used_bytes     = 0;
-   shm->shm_used_bytes_HWM = 0;
```
Comment (gapisback) [inline src/platform_linux/shmem.c:374]: The chunk of deleted inline code is now nicely replaced by a single call to `platform_save_usage_stats()` on L475, below.

File: src/platform_linux/shmem.c [hunk 3]
```diff
@@ -497,53 +505,13 @@ platform_shmdestroy(platform_heap_id *hid_out)
...
-                        ", Large fragments in-use HWM=%u (found in-use=%d)"
-                        ", consumed=%lu bytes (%s)"
-                        ".\n",
-                        shm_nfrees,
-                        shm_nfrees_last_frag,
-                        (FRACTION_ARGS(freed_last_frag_pct) * 100),
-                        nlarge_frags_inuse_HWM,
-                        large_frags_found_in_use,
-                        used_by_large_frags_bytes,
-                        size_str(used_by_large_frags_bytes));
```
Comment (gapisback) [inline src/platform_linux/shmem.c:545]: This messy chunk of deleted code is now all handled cleanly inside expanded fn `platform_shm_print_usage_stats()`, on L514 below.

**Comment by gapisback** [comment]
@rtjohnso - This is the last of the peel-off precursor change sets, paving the ground for reviewing PR #569.

I am assigning this to you, optimistically, expecting that the CI-jobs will all succeed. There is no logic change in core shared memory mgmt itself, so I expect CI-jobs will all succeed.

If you can take a quick look at this, I can merge this immediatelywhen CI passes.

And then will start revamping the large part-3 change-set to prepare PR #569 for review.

**Comment by gapisback** [comment]
Gentle reminder @rtjohnso -- in the hope that this ping via this UI will somehow get through to you.

The CI-jobs have succeeded ... and I'm all ready to check this in. Can you give this a quick look, please?

**Review by rtjohnso — APPROVED** [review]
LGTM

**Pull request merged**